lib/spack/spack/cmd/develop.py

```diff
@@ -84,86 +79,150 @@ def _retrieve_develop_source(spec: spack.spec.Spec, abspath: str) -> None:
-    if not version:
-        # look up the maximum version so infintiy versions are preferred for develop
-        version = max(spec.package_class.versions.keys())
-        tty.msg(f"Defaulting to highest version: {spec.name}@{version}")
-    spec.versions = spack.version.VersionList([version])
+def _update_config(spec, path):
+    find_fn = lambda section: spec.name in section
 
-    # If user does not specify --path, we choose to create a directory in the
-    # active environment's directory, named after the spec
-    path = args.path or spec.name
-    if not os.path.isabs(path):
-        abspath = spack.util.path.canonicalize_path(path, default_wd=env.path)
-    else:
-        abspath = path
+    entry = {"spec": str(spec)}
+    if path != spec.name:
+        entry["path"] = path
 
-    # clone default: only if the path doesn't exist
-    clone = args.clone
-    if clone is None:
-        clone = not os.path.exists(abspath)
+    def change_fn(section):
+        section[spec.name] = entry
 
-    if not clone and not os.path.exists(abspath):
-        raise SpackError("Provided path %s does not exist" % abspath)
+    spack.config.change_or_add("develop", find_fn, change_fn)
 
-    if clone:
-        if os.path.exists(abspath):
-            if args.force:
-                shutil.rmtree(abspath)
-            else:
-                msg = "Path %s already exists and cannot be cloned to." % abspath
-                msg += " Use `spack develop -f` to overwrite."
-                raise SpackError(msg)
 
-        _retrieve_develop_source(spec, abspath)
+def update_env_file(
+    env: spack.environment.Environment,
+    spec: spack.spec.Spec,
+    specified_path: Optional[str] = None,
+    build_dir: Optional[str] = None,
+):
+    """
+    Update the spack.yaml file with additions or changes from a develop call
+    """
+    tty.debug(f"Updating develop config for {env.name} transactionally")
+
+    # this check has to happen outside the write transaction otherwise updates don't seem to happen
+    # correctly
+    if not specified_path:
+        dev_entry = env.dev_specs.get(spec.name)
+        if dev_entry:
+            specified_path = dev_entry.get("path", None)
 
-    tty.debug("Updating develop config for {0} transactionally".format(env.name))
     with env.write_transaction():
-        if args.build_directory is not None:
+        if build_dir is not None:
             spack.config.add(
-                "packages:{}:package_attributes:build_directory:{}".format(
-                    spec.name, args.build_directory
-                ),
+                f"packages:{spec.name}:package_attributes:build_directory:{build_dir}",
                 env.scope_name,
             )
-        _update_config(spec, path)
+        # add develop spec and update path
+        if specified_path:
+            # always overwrite path if it was supplied
+            _update_config(spec, specified_path)
+        else:
+            _update_config(spec, spec.name)
```

Comment:
If the path is just defaulting to the spec name, we don't need to pass it in at all
```suggestion
        # always overwrite path if it was supplied
        _update_config(spec, specified_path)
```